src/Evolution/Systems/Cce/Equations.hpp

```diff
@@ -0,0 +1,724 @@
+ *
+ * The integrand terms which may be computed for a given Bondi variable are
+ * enumerated in the type alias `integrand_terms_to_compute_for_bondi_variable`,
+ * which takes as a single template argument the tag for which integrand terms
+ * would be computed, and is a `tmpl::list` of the integrand terms needed.
+ *
+ * The resulting quantity is returned by `not_null` pointer, and the required
+ * argument tags are given in `return_tags` and `argument_tags` type aliases,
+ * where the `return_tags` are passed by `not_null` pointer (so include
+ * temporary buffers) and the `argument_tags` are passed by const reference.
+ *
+ * Additional mathematical details for each of the computations can be found in
+ * the template specializations of this struct. All of the specializations have
+ * a static `apply` function which takes as arguments
+ * `SpinWeighted<ComplexDataVector, N>`s for each of the Bondi arguments.
+ */
+template <typename IntegrandTag>
+struct ComputeBondiIntegrand;
+
+/*!
+ * \brief Computes the integrand (right-hand side) of the equation which
+ * determines the radial (y) dependence of the Bondi quantity \f$\beta\f$.
+ *
+ * \details The quantity \f$\beta\f$ is defined via the Bondi form of the
+ * metric:
+ * \f[
+ * ds^2 = - \left(e^{2 \beta} (1 + r W) - r^2 h_{AB} U^A U^B\right) du^2 - 2
+ * e^{2 \beta} du dr - 2 r^2 h_{AB} U^B du dx^A + r^2 h_{A B} dx^A dx^B. \f]
+ * Additional quantities \f$J\f$ and \f$K\f$ are defined using a spherical
+ * angular dyad \f$q^A\f$:
+ * \f[ J \equiv h_{A B} q^A q^B, K \equiv h_{A B} q^A
+ * \bar{q}^B.\f]
+ * See \cite Bishop1997ik \cite Handmer2014qha for full details.
+ *
+ * in the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$
+ * is determined by the worldtube initial data, the equation which
+ * determines \f$\beta\f$ on a surface of constant \f$u\f$ given \f$J\f$ on
+ * the same surface is, \f[\partial_y (\beta) = \frac{1}{8} (-1 + y)
+ * \left(\partial_y (J) \partial_y
+ * (\bar{J}) -  \frac{(\partial_y (J \bar{J}))^2}{4 K^2}\right). \f]
+ */
+template <>
+struct ComputeBondiIntegrand<Tags::Integrand<Tags::Beta>> {
+ public:
+  using return_tags = tmpl::append<
+      tmpl::list<Tags::Integrand<Tags::Beta>>,
+      temporary_tags_needed_for_integrand<Tags::Integrand<Tags::Beta>>>;
+  using argument_tags = tmpl::append<
+      pre_swsh_derivative_tags_needed_for_integrand<
+          Tags::Integrand<Tags::Beta>>,
+      swsh_derivative_tags_needed_for_integrand<Tags::Integrand<Tags::Beta>>,
+      integration_independent_tags_needed_for_integrand<
+          Tags::Integrand<Tags::Beta>>>;
+
+  static void apply(
+      const gsl::not_null<SpinWeighted<ComplexDataVector, 0>*>  // NOLINT
+          integrand_for_beta,
+      const SpinWeighted<ComplexDataVector, 2>& dy_j,        // NOLINT
+      const SpinWeighted<ComplexDataVector, 2>& j,           // NOLINT
+      const SpinWeighted<ComplexDataVector, 0>& one_minus_y  // NOLINT
```

Comment:
You have to make sure that the apply function actually take the types of the return and argument tags. Since you want to avoid the many `get`s to bypass the `Scalar`s, you’ll need an `apply` and `apply_impl` or similar, I think @nilsdeppe suggested something.

Reply:
Ah, true. I agree this needs to be fixed.

Reply:
To make sure this works, I think it would be good to add a test that calls `db::mutate_apply`. This will be made easier with #1413